agents/ModularAgent.py

```diff
@@ -0,0 +1,649 @@
+				self.when_learner.ifit(exp.rhs, variablize_by_where(state,exp.mapping.values()), reward)
+			else:
+				self.when_learner.ifit(exp.rhs, state, reward)
+			self.which_learner.ifit(exp.rhs, state, reward)
+			# print("TRAIN ON", list(exp.mapping.values()))
+			self.where_learner.ifit(exp.rhs, list(exp.mapping.values()), state, reward)
+
+
+	def train(self,state, selection, action, inputs, reward, skill_label, foci_of_attention):# -> return None
+		sai = SAIS(selection, action, inputs)
+		state_featurized,knowledge_base = self.apply_featureset(state)
+		# print("TOTAL SKILLS:", len(self.skills))
+		explanations = self.explanations_from_skills(state_featurized,knowledge_base,sai,self.rhs_list)
+		# explanations = [x for x in explanations]
+		# print("SKILL EXPS:", len(explanations))
+		explanations, nonmatching_explanations = self.where_matches(explanations,state_featurized)
+
+		#TODO: DOES THIS EVER HAPPEN???? MAKING IT BREAK BECAUSE I WANT TO SEE IT. (maybe will happen with a different wherelearner)
+		if(len(nonmatching_explanations) > 0):
+			raise ValueError()
+		print("MATCHING:", len(explanations), "NON-MATCHING:", len(nonmatching_explanations), "TOTAL:", len(self.rhs_list))
+		for x in self.rhs_list:
+			print(x.input_rule)
+		# print("WHERE EXPS:", len(explanations))
+
+		if(len(explanations) == 0 ):
+			if(len(nonmatching_explanations) > 0):
+				explanations = [choice(nonmatching_explanations)]
+			else:
+				explanations = self.explanations_from_how_search(state_featurized,sai)
+				explanations = self.which_learner.cull_how(explanations) #Choose all or just the most parsimonious 
+
+				rhs_by_how = self.rhs_by_how.get(skill_label, {})
+				for exp in explanations:
+					if(exp.rhs.as_tuple in rhs_by_how):
+						exp.rhs = rhs_by_how[exp.rhs.as_tuple] 
+					else:
+						rhs_by_how[exp.rhs.as_tuple] = exp.rhs
+						self.rhs_by_how[skill_label] = rhs_by_how
+						self.add_rhs(exp.rhs)
+							
+					
+		# skill = which_learner.most_relevant(skills)
+		self.fit(explanations,state_featurized,reward)
+
+
+	#####--------------------------CHECK-----------------------------###########
+
+	def check(self, state, sai):
+		state_featurized,knowledge_base = self.apply_featureset(state)
+		explanations = self.explanations_from_skills(state,sai,self.rhs_list)
+		explanations, nonmatching_explanations = self.where_matches(explanations)
+		return len(explanations) > 0
+
+
+#####--------------------------CLASS DEFINITIONS-----------------------------###########
+
+
+class SAIS(object):
+	def __init__(self,selection, action, inputs,state=None):
+		self.selection = selection
+		self.action = action
+		self.inputs = inputs# functions that return boolean 
+		self.state = state
+	def __repr__(self):
+		return "S:%r, A:%r, I:%r" %(self.selection,self.action, self.inputs)
+
+
+# class Operator(object):
```

Comment:
Could we get a pass at some dead code removal?

Reply:
https://github.com/cmaclell/apprentice_learner_api/pull/29/commits/638fa053310cd604f6cbe097369129d272566346